RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterMakeLocalCopyQuickFix.cs

```diff
@@ -0,0 +1,182 @@
+﻿using Rubberduck.Inspections.Abstract;
+using System.Linq;
+using Rubberduck.Parsing;
+using Rubberduck.VBEditor;
+using Rubberduck.Inspections.Resources;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Symbols;
+using System.Windows.Forms;
+using Rubberduck.UI.Refactorings;
+using Rubberduck.Common;
+using Antlr4.Runtime;
+using System.Collections.Generic;
+using Antlr4.Runtime.Tree;
+
+namespace Rubberduck.Inspections.QuickFixes
+{
+    public class AssignedByValParameterMakeLocalCopyQuickFix : QuickFixBase
+    {
+        private readonly Declaration _target;
+        private string _localCopyVariableName;
+        private string[] _variableNamesAccessibleToProcedureContext;
+        private bool _isQuickFixUnitTest;
+
+        public AssignedByValParameterMakeLocalCopyQuickFix(Declaration target, QualifiedSelection selection)
+            : base(target.Context, selection, InspectionsUI.AssignedByValParameterMakeLocalCopyQuickFix)
+        {
+            _target = target;
+            _isQuickFixUnitTest = false;
+            _localCopyVariableName = "x" + _target.IdentifierName.CapitalizeFirstLetter();
+            _variableNamesAccessibleToProcedureContext = GetVariableNamesAccessibleToProcedureContext(_target.Context.Parent.Parent);
+        }
+
+        public override bool CanFixInModule { get { return false; } }
+        public override bool CanFixInProject { get { return false; } }
+
+        //This function exists solely to support unit testing 
+        public void TESTONLY_FixUsingAutoGeneratedName(string suggestedName = "")
+        {
+            //Prevent the popup dialog and forces the use of the Suggested Name
+            _isQuickFixUnitTest = true;
+            if (suggestedName.Length > 0)
+            {
+                _localCopyVariableName = suggestedName;  //unit test poses as user input
+            }
+
+            Fix();
+        }
+
+        public override void Fix()
+        {
+            SetLocalCopyVariableName();
+
+            if (!ProposedLocalVariableNameIsValid() || IsCancelled)
+            {
+                return;
+            }
+
+            ReplaceAssignedByValParameterReferences();
+
+            InsertLocalVariableDeclarationAndAssignment();
+        }
+
+        private void SetLocalCopyVariableName()
+        {
+            using (var view = new AssignedByValParameterQuickFixDialog(_target, Selection))
+            {
+                view.NewName = _localCopyVariableName;
+                view.IdentifierNamesAlreadyDeclared = _variableNamesAccessibleToProcedureContext;
+                if (!_isQuickFixUnitTest)
```

Comment:
The problem is that there's this `AssignedByValParameterQuickFixDialog` dependency that needs to be tamed.

You could constructor-inject an abstract factory here - and Ninject will do all the hard work!

    public interface IAssignedByValParameterQuickFixDialogFactory
    {
        IAssignedByValParameterQuickFixDialog Create(Declaration target, Selection selection, string[] accessibleNames);
    }

Just make the dialog implement an interface that's named after it (with the `I` prefix) - that interface could look like this:

    public interface IAssignedByValParameterQuickFixDialog : IDisposable
    {
        void ShowDialog();
        DialogResult DialogResult { get; }
        string NewName { get; }
    }

Make the dialog implement that interface (just tacking `: IAssignedByValParameterQuickFixDialog` to its definition should be enough, from what I'm seeing)

And then the `AssignedByValParameterMakeLocalCopyQuickFix` can take an `IAssignedByValParameterQuickFixDialogFactory` constructor parameter and store that abstract factory in a private readonly field:

    private readonly IAssignedByValParameterQuickFixDialogFactory _dialogFactory;

Now, instead of doing `using (var view = new AssignedByValParameterQuickFixDialogFactory(_target, Selection))`, you can do this:

    using (var view = _dialogFactory.Create(_target, Selection, _variableNamesAccessibleToProcedureContext))

And now whoever creates a `AssignedByValParameterMakeLocalCopyQuickFix` is responsible for passing the dependencies to the constructor - I think that would be the inspection result class... so **for now** let's make that inspection result's constructor intake the `IAssignedByValParameterQuickFixDialogFactory` dependency through its constructor, and then since the inspection result is `new`'d up by the inspection itself, the inspection's constructor **for now** will have to intake the `IAssignedByValParameterQuickFixDialogFactory` dependency through its constructor as well.

That's dirty dependency injection, but it should "just work". We'll revisit the *injecting depedencies of dependencies of dependencies* issue when the inspections feature is re-architected.